### PR TITLE
Update c bindings, add new functions

### PR DESCRIPTION
**Pull request by DyrellC**
# Description of change
Updating c bindings to include tagged and signed packets on the subscriber end, as well as a couple utility functions for usability (like converting string link representation to proper streams Address object through bindings).

## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested
I have run each function through modifications of the main.c file, and confirmed they work when pulled into golang

## Change checklist
- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: bindings/c/src/api/mod.rs [hunk 1]
```diff
@@ -23,6 +23,24 @@ use cty::{
 
 use iota::client::bytes_to_trytes;
 
+#[no_mangle]
+pub extern "C" fn address_from_string(c_addr: *const c_char) -> *const Address {
+    unsafe {
+        let addr_vec: Vec<&str> = CStr::from_ptr(c_addr).to_str().unwrap().split(":").collect();
+        let addr = Address::from_str(addr_vec[0], addr_vec[1]).unwrap();
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:30]: `addr_vec[1]` will panic if address doesn't contain a colon, need to add a guard. 

`unwrap()` in general can panic, need better handling of unwraps throughout all streams repo, not only bindings. In c bindings we can return nulls.

File: bindings/c/src/api/mod.rs [hunk 2]
```diff
@@ -136,6 +154,11 @@ impl Default for MessageLinks {
     }
 }
 
+#[no_mangle]
+pub extern "C" fn new_message_links(msg_link: *const Address, seq_link: *const Address) -> *mut MessageLinks {
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:158]: No need to make an explicit function, `message_links_t` struct is defined in `channels.h` and can be constructed directly in c.

File: bindings/c/src/api/mod.rs [hunk 3]
```diff
@@ -269,19 +292,34 @@ pub extern "C" fn get_msgid_str(msgid: *mut MsgId) -> *const c_char {
...
     unsafe {
         CString::new(bytes_to_trytes((*address).msgid.as_ref()).to_string()).unwrap().into_raw()
     }
 }
 
+
+#[no_mangle]
+pub extern "C" fn get_address_id_str(address: *mut Address) -> *mut c_char {
+    unsafe {
+        CString::new(hex::encode((*address).msgid.as_ref())).unwrap().into_raw()
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:319]: Shall we check pointers for `null`, or hope/expect caller code doesn't violate convention?

**Review by semenov-vladyslav — CHANGES_REQUESTED** [review]
Update `main.c` to cover the new functions.

File: bindings/c/CMakeLists.txt [hunk 4]
```diff
@@ -5,7 +5,7 @@ include(FetchContent)
 
 option(NO_STD "Enable no_std build, without iota_client" OFF)
 option(SYNC_CLIENT "Enable sync transport via iota_client" ON)
-option(STATIC "Build shared library" ON)
+option(STATIC "Build shared library" OFF)
```
Comment (kwek20) [inline bindings/c/CMakeLists.txt:8]: ahyes we should rename this and maybe decide on what we make as default, we keep swapping it aboot

File: bindings/c/src/api/mod.rs (hunk 3 above)
Comment (DyrellC) [reply bindings/c/src/api/mod.rs:303]: Hmmm a good question really. I think it's fair to expect that they'll not violate the conventions, but it probably doesn't hurt to catch the strays and save them the seg fault

File: bindings/c/src/api/mod.rs (hunk 1 above)
Comment (DyrellC) [reply bindings/c/src/api/mod.rs:29]: Will map_or them to ensure null returns wherever a panic could occur

File: bindings/c/CMakeLists.txt (hunk 4 above)
Comment (DyrellC) [reply bindings/c/CMakeLists.txt:8]: We should default to Shared since static is still giving us grief

**Review by kwek20 — APPROVED** [review]
LGTM!

File: bindings/c/src/api/mod.rs [hunk 5]
```diff
@@ -21,7 +21,29 @@ use cty::{
     uint8_t,
 };
 
-use iota::client::bytes_to_trytes;
+
+#[no_mangle]
+pub extern "C" fn address_from_string(c_addr: *const c_char) -> *const Address {
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:26]: This is a nice function in general, it's going to be used in other bindings. Maybe move the impl to the main crate?

**Review by semenov-vladyslav — CHANGES_REQUESTED** [review]
Looks good, but move impl of `address_from_string` into main crate?

File: bindings/c/src/api/mod.rs (hunk 5 above)
Comment (DyrellC) [reply bindings/c/src/api/mod.rs:26]: Yeah I think that's a good idea, i'll do that before push

**Pull request merged**